File: Data Structures & Algorithms/palindrome-partitioning/submission-5.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> curr;
        dfs(s, res, curr, 0);
        return res;
    }
    void dfs(string s, vector<vector<string>>&res, vector<string>&curr, int i){
        if(i>=s.length()){
            res.push_back(curr);
            return;
        }
        for(int j = i; j < s.length(); j++){
            if(ispali(s, i, j)){
                curr.push_back(s.substr(i, j-i+1));
                dfs(s, res, curr, j+1);
                curr.pop_back();
            }
        }
    }
    bool ispali(string s, int i, int j){
        while(i<=j){
            if(s[i]==s[j]){
                i++;
                j--;
            }
            else{
                return false;
            }
        }
        return true;
    }
};

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-5.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mp = {
            '2' : 'abc',
            '3': "def",
            '4': "ghi",
            '5':"jkl",
            '6':'mno',
            '7':"pqrs",
            '8':"tuv",
            '9':"wxyz",
        }
        if not digits:
            return []
        res, curr = [], []
        def dfs(i):
            if len(digits) == i:
                copy = ''.join(curr)
                res.append(copy)
                return
            for j in mp[(digits[i])]:
                curr.append(j)
                dfs(i+1)
                curr.pop()

        dfs(0)
        return res
